### PR TITLE
uniform and symmetric quantizers implementation

### PR DESCRIPTION
**Pull request by ofirgo**

### CODE REVIEW

File: model_compression_toolkit/common/quantization/quantization_params_generation/symmetric_selection.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+    Returns:
+        Optimal threshold to quantize the histogram a symmetric manner.
+    """
+    tensor_max = np.max(np.abs(bins))
+    signed = np.any(bins < 0)  # check if tensor is singed
+    if threshold_method == qc.ThresholdSelectionMethod.NOCLIPPING:
+        return {THRESHOLD: tensor_max}
+    else:
+        # TODO: what happens with z-filter?
```
Comment (reuvenperetz) [inline model_compression_toolkit/common/quantization/quantization_params_generation/symmetric_selection.py]: Let's check

File: tests/keras_tests/function_tests/test_uniform_range_selection_weights.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        quantized_model, user_info = _quantize_fixed_bit_widths_graph(False,
+                                                                      fw_info,
+                                                                      None,
+                                                                      lambda: [np.random.randn(1, 16, 16, 4)],
+                                                                      None,
+                                                                      tg,
+                                                                      keras_impl)
+
+        nodes_list = list(graph.nodes)
+        conv1_min = nodes_list[0].candidates_weights_quantization_cfg[0].weights_quantization_params['range_min'].flatten()
```
Comment (reuvenperetz) [inline tests/keras_tests/function_tests/test_uniform_range_selection_weights.py]: use strings from constants

File: model_compression_toolkit/common/quantization/quantization_params_generation/qparams_activations_computation.py [hunk 3]
```diff
@@ -50,7 +51,8 @@ def get_activations_qparams(n: BaseNode,
     else:
         signed = np.any(bins_values < 0)
 
-    if n.prior_info.is_output_bounded():
+    if n.prior_info.is_output_bounded() and \
+            n.activation_quantization_cfg.activation_quantization_method == QuantizationMethod.POWER_OF_TWO:
         n.activation_quantization_cfg.activation_quantization_params_fn = quantization_params_generation.no_clipping_selection_min_max
```
Comment (reuvenperetz) [inline model_compression_toolkit/common/quantization/quantization_params_generation/qparams_activations_computation.py]: In the case of symmetric/uniform, do we use the collected statistics even when the node has a bounded output?

File: model_compression_toolkit/common/quantization/quantization_params_generation/symmetric_selection.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        n_iter: Number of iterations to search for the optimal threshold (not used for this method).
+        min_threshold: Minimal threshold to use if threshold is too small (not used for this method).
+        quant_error_method: an error function to optimize the parameters' selection accordingly.
+
+    Returns:
+        Optimal threshold to quantize the tensor in a symmetric manner.
+    """
+
+    signed = np.any(tensor_data < 0)  # check if tensor is singed
+    unsigned_tensor_data = np.abs(tensor_data)
```
Comment (reuvenperetz) [inline model_compression_toolkit/common/quantization/quantization_params_generation/symmetric_selection.py:66]: Maybe we shouldn't check the sign, as it's always signed in weights quantization

File: model_compression_toolkit/common/quantization/quantization_params_fn_selection.py [hunk 5]
```diff
@@ -16,86 +16,98 @@
...
             params_fn = None
+    elif activation_quantization_method == QuantizationMethod.SYMMETRIC:
+        params_fn = symmetric_selection_histogram
+    elif activation_quantization_method == QuantizationMethod.UNIFORM:
+        params_fn = uniform_selection_histogram
     else:
         raise Exception(
             f'No params function for the configuration of quantization method {activation_quantization_method} and '
-            f'threshold selection method {activation_threshold_method}')
+            f'threshold selection method {activation_error_method}')
```
Comment (haihabi) [inline model_compression_toolkit/common/quantization/quantization_params_fn_selection.py]: threshold selection should be change to quantization error

File: model_compression_toolkit/common/quantization/quantization_params_generation/kl_selection.py [hunk 6]
```diff
@@ -385,3 +395,36 @@ def _get_sliced_histogram(bins: np.ndarray,
...
+    Vectorized implementation for the KL error function,
+    to allow vectorized per-channel computation (for weights quantization).
+    Input tensor x is expected to be reshaped before passed to this method,
+    such that axis 0 contains the number of channels and the tensor values are placed along axis 1.
+    Note that the quantized value of x is not directly used in this method,
+    but rather being calculated during the process of gathering the distribution statistics
+    within the _kl_error_function.
+
+    Args:
+        x:
```
Comment (haihabi) [inline model_compression_toolkit/common/quantization/quantization_params_generation/kl_selection.py]: Args comments are missing:
x is a ndarray of shape n x m where n is the number of channels and m is the number of elements in each channel

File: model_compression_toolkit/common/quantization/quantization_params_generation/qparams_activations_computation.py (hunk 3 above)
Comment (haihabi) [inline model_compression_toolkit/common/quantization/quantization_params_generation/qparams_activations_computation.py]: I agree with @reuvenperetz  in the case of bounded output. We can use this information to set the threshold or min-max values.

File: model_compression_toolkit/common/similarity_analyzer.py [hunk 7]
```diff
@@ -53,19 +53,23 @@ def tensor_norm(x: np.ndarray, p: float = 2.0) -> np.float:
...
+        norm: whether to normalize the error function result.
 
     Returns:
         The MSE distance between the two tensors.
     """
     validate_before_compute_similarity(float_tensor, fxp_tensor)
-    return np.power(float_tensor - fxp_tensor, 2.0).mean()
+    error = np.power(float_tensor - fxp_tensor, 2.0).mean()
+    if norm:
+        error /= (np.power(float_tensor, 2.0).mean() + 1e-8)
```
Comment (haihabi) [inline model_compression_toolkit/common/similarity_analyzer.py]: Make the eps value a function parameter with a default value 1e-8, this is true for all metrics

File: requirements.txt [hunk 8]
```diff
@@ -1,10 +1,15 @@
 networkx==2.5
```
Comment (haihabi) [inline requirements.txt]: Why did you change the requirements?

Comment (ofirgo) [reply requirements.txt]: By accident, the pycharm showed me a message that the requirements file doesn't contain all the necessary requirements and I chose the option of solving it automatically.
I'll revert those changes, but nonetheless, shouldn't I include scipy in the requirements now? because I'm using it for the minimization.

File: model_compression_toolkit/common/quantization/quantization_params_generation/symmetric_selection.py (hunk 4 above)
Comment (ofirgo) [reply model_compression_toolkit/common/quantization/quantization_params_generation/symmetric_selection.py:66]: I followed the same technique that is used for power-of-two threshold selection, you can see in the method [qparams_selection_tensor_search](https://github.com/ofirgo/model_optimization/blob/6c09e77ee9c17a04577d4162b87d5d5ae6420bec/model_compression_toolkit/common/quantization/quantization_params_generation/qparams_search.py#L55) that it also checks signed value and passes it to the [quantize_tensor](https://github.com/ofirgo/model_optimization/blob/6c09e77ee9c17a04577d4162b87d5d5ae6420bec/model_compression_toolkit/common/quantization/quantization_params_generation/qparams_search.py#L84) method.
Is it a bug? if so, I'll change this is my implementation and in the original.

Comment (reuvenperetz) [reply model_compression_toolkit/common/quantization/quantization_params_generation/symmetric_selection.py:66]: Yeah... I'm afraid it's a bug.
@haihabi What do you think? We should pass signed=True always in the case of weights quantization, right?

File: requirements.txt [hunk 9]
```diff
@@ -1,4 +1,3 @@
-networkx==2.5
```
Comment (haihabi) [inline requirements.txt]: Why the networkx is removed?

File: model_compression_toolkit/common/quantization/quantization_params_generation/symmetric_selection.py (hunk 4 above)
Comment (ofirgo) [reply model_compression_toolkit/common/quantization/quantization_params_generation/symmetric_selection.py:66]: Resolved offline

File: requirements.txt (hunk 9 above)
Comment (ofirgo) [reply requirements.txt]: fixed
